module/Finna/src/Finna/OrganisationInfo/Provider/Kirkanta.php

```diff
@@ -0,0 +1,644 @@
+                    $base = $this->config->General[$url];
+                    $conf = ['base' => $base];
+
+                    if (preg_match_all('/{([^}]*)}/', $base, $matches)) {
+                        $conf['params'] = $matches[1];
+                    }
+                    $mapUrlConf[$url] = $conf;
+                }
+            }
+            foreach ($mapUrlConf as $map => $mapConf) {
+                $mapUrl = $mapConf['base'];
+                $replace = [];
+                foreach ($mapConf['params'] ?? [] as $param) {
+                    if ($val = $response['address'][$param] ?? null) {
+                        $replace[$param] = $val;
+                    }
+                }
+                foreach ($replace as $param => $val) {
+                    $mapUrl = str_replace(
+                        '{' . $param . '}',
+                        rawurlencode($val),
+                        $mapUrl
+                    );
+                }
+                $result[$map] = $mapUrl;
+            }
+        } else {
+            foreach ($mapUrls as $url) {
+                $result[$url] = null;
+            }
+        }
+        $schedules = [
+            'schedule' => $response['schedules'],
+            'status' => $response['liveStatus'],
+        ];
+        $result['openTimes'] = $this->parseSchedules($schedules);
+        $result['openNow'] = $result['openTimes']['openNow'];
+
+        $phones = [];
+        foreach ($response['phoneNumbers'] ?? [] as $phone) {
+            // Check for email data in phone numbers
+            if (str_contains($phone['number'], '@')) {
+                continue;
+            }
+            if ($name = $phone['name']) {
+                $phones[] = [
+                    'name' => $name,
+                    'number' => $phone['number'],
+                ];
+            }
+        }
+        $result['phones'] = $phones;
+
+        $emails = [];
+        $dedupEmails = array_unique($response['emailAddresses'] ?? [], SORT_REGULAR);
+        foreach ($dedupEmails as $address) {
+            $emails[] = [
+                'name' => $address['name'],
+                'email' => $address['email'],
+            ];
+        }
+        $result['emails'] = $emails;
+
+        $pics = [];
+        foreach ($response['pictures'] ?? [] as $pic) {
+            $pics[] = $pic['files']['medium'];
+        }
+        $result['pictures'] = $pics;
+
+        $links = [];
+        foreach ($response['links'] ?? [] as $link) {
+            $name = $link['name'];
+            $url = $link['url'];
+            if ($name && $url) {
+                $links[] = ['name' => $name, 'url' => $url];
+            }
+        }
+        $result['links'] = $links;
+
+        $result['services'] = [];
+        $result['allServices'] = [];
+        if (!empty($response['services'])) {
+            $servicesMap = [];
+            $servicesConf = $this->config->OpeningTimesWidget->services->toArray();
```

Comment:
Tän vois varmaan siirtää ylöspäin ja sit vaikka:

$servicesConf = $this->config->OpeningTimesWidget?->services?->toArray() ?? [];

Reply:
Miksi ylemmäs? Heivasin koko $servicesConf:n.

Reply:
Eikun oha se iha fine tossa, joku sulku tuli. Mut ilman pohja-asetuksia niin kaatuu.

Reply:
Laitoin kysymysmerkit juu.